juju/juju
# Validate api client imports

To ensure that we don't accidentally bring in an import for the api client packages, I've added a static analysis to validate that.

In PR[1] it was noticed that we brought in the whole of the aws library. That wasn't intentional, so we need to prevent ourselves from doing it again. This should hopefully go some way to do that.

 1. https://github.com/juju/juju/pull/16633

## Checklist


## QA steps


```sh
$ make static-analysis
```


If you apply this without rebasing, you'll see the following error:

```sh
➜ make static-analysis
dqlite already installed

==> Checking for dependencies
==> Using Juju located at /home/simon/go/bin/juju
==> TEST BEGIN: test_static_analysis (tmp.Tnh)
===> [ ✔ ] Success: copyright (1s)
===> [ ✔ ] Success: check doc go (10s)
===> [ ✔ ] Success: api client (9s)
===> [ x ] Fail: go (1s)
==> Cleaning up
====> Completed cleaning up pids
====> Completed cleaning up jujus

==> TESTS DONE: test_static_analysis
==> RUN OUTPUT: test_static_analysis
    | Checking ./api/client/action
    | Checking ./api/client/annotations
    | Checking ./api/client/application
    | Checking ./api/client/applicationoffers
    | Checking ./api/client/backups
    | Checking ./api/client/block
    | Checking ./api/client/bundle
    | Checking ./api/client/charms
    | Import not allowed: github.com/aws/aws-sdk-go-v2
    | Consult the list of allowed imports.
    | Error: API Client import failure in ./api/client/charms
    | Checking ./api/client/client
    | Checking ./api/client/cloud
    | Checking ./api/client/credentialmanager
    | Checking ./api/client/highavailability
    | Checking ./api/client/imagemetadatamanager
    | Checking ./api/client/keymanager
    | Checking ./api/client/machinemanager
    | Checking ./api/client/metricsdebug
    | Checking ./api/client/modelconfig
    | Checking ./api/client/modelgeneration
    | Checking ./api/client/modelmanager
    | Checking ./api/client/modelupgrader
    | Checking ./api/client/payloads
    | Checking ./api/client/resources
    | Checking ./api/client/secretbackends
    | Checking ./api/client/secrets
    | Checking ./api/client/spaces
    | Checking ./api/client/sshclient
    | Checking ./api/client/storage
    | Checking ./api/client/subnets
    | Checking ./api/client/usermanager
```

## Links

**Jira card:** JUJU-5085
